docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,783 @@
+dependency.*
+
+## Failed Tasks
+
+No special handling of failed tasks is required.
+- if success is required, a failed task automatically gets retained in the task
+  pool as incomplete.
+- if success is optional (which would usually imply that optional failure is
+  explicitly handled), a failed task can be removed as complete
+
+No special syntax (e.g. `c:fail => NULL`) or suicide triggers (e.g. `c:fail =>
+!c`) are needed for leaf tasks either, if we don't care if they succeed or fail:
+```
+a => b => c?  # success of c is optional
+```
+
+## Family Triggers
+
+**NOTE IMPLICATIONS OF FAMILY TRIGGERS STILL UNDER DISCUSSOIN**
+
+Conceivably we could try to interpret the output signifier in terms of a
+collective pseudo-output for the family as whole, but that gets really hard to
+understand.
+
+Fortunately, the output type signifier can simply be applied directly to all
+family members, for all family trigger types. That's easy to understand,
+and it makes sense because the signifier doesn't affect triggering, it only
+affects what we do with family members if the outputs in question are not
+completed (i.e. keep them in n=0 as incomplete, or forget them).
+
+If particular member outputs need special treatment they should be singled
+out in the graph - just like special member triggers.
+
+```
+A:succeed-all => b
+```
+- trigger b only if all members succeed; and all members are required to
+  succeed
+
+```
+A:succeed-all? => b
+```
+- trigger b only if all members succeed; but it's OK if they fail (maybe the
+  failed path is handled elsewhere)
+
+```
+A:succeed-any => b
+```
+- trigger b if any member succeeds; but any member that runs is required to
+  succeed
+
+```
+A:succeed-any? => b
+```
+- trigger b if any member succeeds; but it's OK if any or all members fail
+
+
+## Backward Compatibility (Cylc 7)
+
+In Cylc 7 graphs, all outputs are effectively "required" in that all tasks are
+pre-spawned to the next cycle point, and users are expected to define suicide
+triggers to remove unused branches that are really optional.
+
+In Cylc 8, suicide triggers mostly aren't needed because unused branches don't
+get spawned at all, but they're still there and still if needed.
+
+So if a Cylc 7 workflow is detected:
+- instead of flagging use of both `foo:succeed` and `foo:fail` as an error,
+we can infer that success must be optional for `foo`, to avoid ending up
+with an incomplete task due to the unused output
+- similarly for `foo:submit` and `foo:submit-fail`
+- for other outputs, e.g. `foo:x` and `foo:y` we can't know if they're optional
+  or not, so assume they are required and let the existing Cylc 7 suicide
+triggers clean up any resulting incomplete tasks 
+
+### How to detect Cylc 7 workflows?
+
+- detect `suite.rc` vs `flow.cylc` config filename?
+  - users might change the filename without upgrading the syntax?
+- require users to add an explicit marker comment in the file?
+- other?
```

Comment:
If users change the filename without upgrading they will get validation errors for any problems in the graph so will soon realise their mistake. I don't really see why we need to allow users to be able to create `flow.cylc` files with Cylc 7 graphs.

Reply:
Ok I’m happy with that 👍🏻